pypiserver/config.py

```diff
@@ -0,0 +1,664 @@
+
+To add a config option:
+
+- If it should be available for all subcommands (run, update), add it to
+  the `global_args` parser.
+- If it should only be available for the `run` command, add it to the
+  `run_parser`.
+- If it should only be available for the `update` command, add it to the
+  `update_parser`.
+- Add it to the appropriate Config class, `_ConfigCommon` for global options,
+  `RunConfig` for `run` options, and `UpdateConfig` for `update` options.
+- Ensure your config option is tested in `tests/test_config.py`.
+
+The `Config` class provides a `.from_args()` static method, which returns
+either a `RunConfig` or an `UpdateConfig`, depending on which subcommand
+is specified in the args.
+
+Legacy commandline arguments did not require a subcommand. This form is
+still supported, but deprecated. A warning is printing to stderr if
+the legacy commandline format is used.
+
+Command line arguments should be parsed as early as possible, using
+custom functions like the `auth_*` functions below if needed. For example,
+if an option were to take JSON as an argument, that JSON should be parsed
+into a dict by the argument parser.
+"""
+
+import argparse
+import contextlib
+import hashlib
+import io
+import pkg_resources
+import re
+import textwrap
+import sys
+import typing as t
+
+from pypiserver import __version__
+
+
+# Specify defaults here so that we can use them in tests &c. and not need
+# to update things in multiple places if a default changes.
+class DEFAULTS:
+    """Config defaults."""
+
+    AUTHENTICATE = ["update"]
+    FALLBACK_URL = "https://pypi.org/simple/"
+    HASH_ALGO = "md5"
+    INTERFACE = "0.0.0.0"
+    LOG_FRMT = "%(asctime)s|%(name)s|%(levelname)s|%(thread)d|%(message)s"
+    LOG_ERR_FRMT = "%(body)s: %(exception)s \n%(traceback)s"
+    LOG_REQ_FRMT = "%(bottle.request)s"
+    LOG_RES_FRMT = "%(status)s"
+    LOG_STREAM = sys.stdout
+    PACKAGE_DIRECTORIES = ["~/packages"]
+    PORT = 8080
+    SERVER_METHOD = "auto"
+
+
+def auth_arg(arg: str) -> t.List[str]:
+    """Parse the authentication argument."""
+    # Split on commas, remove duplicates, remove whitespace, ensure lowercase.
+    # Sort so that they'll have a consistent ordering.
+    items = sorted(list(set(i.strip().lower() for i in arg.split(","))))
+    # Throw for any invalid options
+    if any(i not in ("download", "list", "update", ".") for i in items):
+        raise ValueError(
+            "Invalid authentication option. Valid values are download, list, "
+            "and update, or . (for no authentication)."
+        )
+    # The "no authentication" option must be specified in isolation.
+    if "." in items and len(items) > 1:
+        raise ValueError(
+            "Invalid authentication options. `.` (no authentication) "
+            "must be specified alone."
+        )
+    return items
+
+
+def hash_algo_arg(arg: str) -> t.Callable:
+    """Parse a hash algorithm from the string."""
+    if arg not in hashlib.algorithms_available:
```

Comment:
Question here: it looks like disabling the hash_algo doesnt work anymore, or am I missing something?

Reply:
Ah, thanks for catching this. I forgot to handle the "falsey" cases. I'll add that in.

Reply:
Added in #347